src/docs/arch-tiered-storage.dox

```diff
@@ -0,0 +1,439 @@
+or an abstraction of a cloud provider with operations. WiredTiger has a
+several instances of \c WT_STORAGE_SOURCE, these include the drivers
+for the AWS, GCP, and Azure clouds.  A storage source can be asked to
+create a custom file system (returning a \c WT_FILE_SYSTEM) from a
+bucket name and credentials.  A file system created this way is stored
+in the \c WT_BUCKET_STORAGE.
+
+The file system can be used for various readonly operations, like
+listing the bucket contents or opening a \c WT_FILE_HANDLE .  That, in
+turn, can be used to get the contents of a cloud object.  Note that \c
+WT_FILE_SYSTEM and \c WT_FILE_HANDLE are more generic WiredTiger
+concepts, and are used outside of tiered storage.
+
+The file system obtained from a storage source cannot be used to write
+pieces of data to objects, rather there is a method on the \c
+WT_STORAGE_SOURCE (cloud driver) that is used to copy a source file in
+its entirety to the cloud. This makes it clear that objects must be
+written in their entirety, but may be read, if desired, in pieces.
+
+In addition to the cloud providers, there is a storage source called
+\c dirstore that emulates the behavior of a cloud provider, but stores
+objects in a directory. This is used for testing, avoiding the
+complication and expense of using cloud storage.  Typically we use a
+subdirectory of the WiredTiger home directory as our dirstore
+repository. When a test breaks, we already have the "cloud" objects
+available for debugging.
+
+We anticipate that buckets may be shared among multiple nodes, and
+possibly multiple clusters. To avoid name collisions, we provide a
+prefix that can be given on a ::wiredtiger_open call, and assume the
+caller gives us a unique prefix. This prefix is stored as
+part of the \c WT_BUCKET_STORAGE, and so can also be
+specified per table upon creation.  The prefix is prepended to every
+name stored in a bucket.
+
+@section ts_flush flush_tier
+
+A WT_SESSION::checkpoint call with \c flush_tier enabled (also known
+as \c flush_tier) puts everything we discussed on this page together.
+It identifies any tiered btrees that have changes since the previous
+checkpoint, and for each such modified btree:
+    -# eviction is temporarily disabled on this btree, while waiting
+       for any writes to the active file to drain
+    -# update the metadata to know about the next object
+    -# a new empty file is created named with the next object number,
+       this will become the table's active file
+    -# switch the table's active file
+    -# add a FLUSH entry to the work queue for the previously active
+       file to be copied to the cloud
+    -# enable eviction on this btree
+
+When the FLUSH work entry is completed, we:
+    -# tell the chunk cache to ingest the object before it is removed
+    -# update the metadata to reference the new object in the cloud
+    -# queue a FLUSH_FINISH operation
+    -# queue a REMOVE_LOCAL operation
+
+@section ts_future Future
```

Comment:
I stopped here. I'm not sure how useful it is to have all the future stuff listed here. It is years old and would likely change if/when it ever gets resuscitated.